sql: factor expression semantic analysis.

knz opened the pull request:
This patch introduces `planner.analyzeExpr` which groups sequences of
calls to `replaceSuqueries`, `resolveQNames`, `TypeCheck` and
`NormalizeExpr`.

Fixes #6860.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7049)

<!-- Reviewable:end -->

knz committed:
sql: factor expression semantic analysis.

This patch introduces `planner.analyzeExpr` which groups sequences of
calls to `replaceSuqueries`, `resolveQNames`, `TypeCheck` and
`NormalizeExpr`.

nvanbenschoten commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[sql/analyze.go, line 1571 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaEXDDuPAtYQpyviy-:-KJaEXDEG6QprogWBIU8:1562745520) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1571)):*
> ```Go
> // - type checking (with optional type enforcement);
> // - normalization.
> // The parameters tables and qvals, if non-nil, indicate qname resolution
> ```

`s/if non-nil/if both are non-nil/`

---

*[sql/analyze.go, line 1574 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaE5No4mCSHeAL-FXj:-KJaE5No4mCSHeAL-FXk:708274760) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1574)):*
> ```Go
> // should be performed. The qvals map will be filled as a result.
> func (p *planner) analyzeExpr(
> 	raw parser.Expr,
> ```

I see why you grouped certain params together, but these parameters don't fit our style guide.

---

*[sql/analyze.go, line 1576 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaEyL4GsvqoXLMlXny:-KJaEyL4GsvqoXLMlXnz:-819099382) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1576)):*
> ```Go
> 	raw parser.Expr,
> 	tables []*tableInfo, qvals qvalMap,
> 	expectedType parser.Datum, requireType bool, expectedContext string,
> ```

`expectedContext` is pretty unclear.

---

*[sql/analyze.go, line 1579 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaFFtGTVzcZZyGNY4t:-KJaFFtGTVzcZZyGNY4u:-396282545) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1579)):*
> ```Go
> ) (parser.TypedExpr, error) {
> 	// Replace the sub-queries.
> 	replaced, err := p.replaceSubqueries(raw, 1)
> ```

Should we be hardcoding in this `1`?

---

*[sql/group.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaFlp4wZrdRUMgufim:-KJaFlp5isrIGew5hU2G:2004508335) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a08cd9a62ebca48f7add95fafe877f1c06ab8ec/sql/group.go#L79)):*
> ```Go
> 		}
> 
> 		// We could potentially skip this, since it will be checked in addRender,
> ```

Looks like we dropped this comment. We might want to keep it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 6 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/analyze.go, line 1571 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaEXDDuPAtYQpyviy-:-KJaoJXlBNlTXNHB6oFz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1571)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `s/if non-nil/if both are non-nil/`
</details>
Done.

---

*[sql/analyze.go, line 1574 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaE5No4mCSHeAL-FXj:-KJaocS75rldPLvgo1A2:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1574)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> I see why you grouped certain params together, but these parameters don't fit our style guide.
</details>
Done.

---

*[sql/analyze.go, line 1576 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaEyL4GsvqoXLMlXny:-KJaod5mN2dQESUgpC3p:-293160254) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1576)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> `expectedContext` is pretty unclear.
</details>
Used `typingContext` instead.

---

*[sql/analyze.go, line 1579 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaFFtGTVzcZZyGNY4t:-KJapYm-BBAk2au9daaM:-586742064) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6b14528452fbe1ba937eb0a9d3ed9e178172dd/sql/analyze.go#L1579)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Should we be hardcoding in this `1`?
</details>
Yeah. I added a comment to clarify.

---

*[sql/group.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-KJaFlp4wZrdRUMgufim:-KJb-xstFl5KgSLsFVv2:-472175298) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a08cd9a62ebca48f7add95fafe877f1c06ab8ec/sql/group.go#L79)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Looks like we dropped this comment. We might want to keep it.
</details>
Or actually even simplify. I don't like doing the work two times only for the sake of errors, which should be (under normal circumstances) the exception, not the norm.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049)*
<!-- Sent from Reviewable.io -->

knz commented:
cc @dt @paperstreet

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> cc @dt @paperstreet 
</details>


---

Reviewed 5 of 8 files at r2, 3 of 3 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7049#-:-KJgA-ws4U3P9BYjwwmd:1417088721)*
<!-- Sent from Reviewable.io -->
